File: kotlin/sdk/src/commonMain/kotlin/com/basecamp/hey/Cache.kt
package com.basecamp.hey

import io.ktor.http.Headers
import io.ktor.http.headers

/**
 * A response the cache holds for a URL: the validator HEY sent with it, the body it
 * validates, and the headers that came with the body, so a 304 answers the page as it was
 * first read — `Link` and `X-Total-Count` included, which a 304 need not repeat.
 */
class CachedResponse(
    /** The `ETag` HEY sent with the body, sent back as `If-None-Match` on the next read. */
    val etag: String,
    /** The body as HEY answered it. */
    val body: ByteArray,
    /** The headers HEY answered the body with, less any credential. */
    val headers: Map<String, List<String>> = emptyMap(),
) {
    /** The cached headers, with each one the 304 carried replacing what was held under its name. */
    internal fun headersUpdatedBy(fresh: Headers): Headers = headers {
        for ((name, values) in this@CachedResponse.headers) appendAll(name, values)
        fresh.forEach { name, values ->
            remove(name)
            appendAll(name, values)
        }
    }

    override fun equals(other: Any?): Boolean =
        other is CachedResponse && other.etag == etag && other.body.contentEquals(body) && other.headers == headers

    override fun hashCode(): Int = 31 * (31 * etag.hashCode() + body.contentHashCode()) + headers.hashCode()

    override fun toString(): String = "CachedResponse(etag=$etag, ${body.size} bytes, ${headers.size} headers)"
}

/** Stores JSON responses by `ETag` so a repeated read can be answered from a 304. */
interface ResponseCache {
    /** What is held under [key], if anything. */
    fun get(key: String): CachedResponse?

    /** Holds a response under [key], replacing whatever was there. */
    fun set(key: String, response: CachedResponse)

    /** Forgets what is held under [key]. */
    fun invalidate(key: String)

    /** Forgets everything. */
    fun clear()
}

/** Keeps cached responses for the life of the process. What `enableCache` turns on. */
class InMemoryCache : ResponseCache {
    private val entries: MutableMap<String, CachedResponse> = createServiceCache()

    override fun get(key: String): CachedResponse? = entries[key]

    override fun set(key: String, response: CachedResponse) {
        entries[key] = response
    }

    override fun invalidate(key: String) {
        entries.remove(key)
    }

    override fun clear() {
        entries.clear()
    }

    /** How many responses are held. */
    val size: Int get() = entries.size
}

/**
 * The key a read is cached under: the URL and the credential it went out with, hashed, so
 * one identity's reading is never answered to another and the token never sits in a key.
 */
internal fun cacheKey(url: String, credential: String): String = sha256Hex("$credential\n$url".encodeToByteArray())

File: swift/Sources/Hey/Cache.swift
import Foundation

/// A response the cache holds for a URL: the validator HEY sent with it, the body it
/// validates, and the headers that came with the body, so a 304 answers the page as it was
/// first read — `Link` and `X-Total-Count` included, which a 304 need not repeat.
public struct CachedResponse: Sendable, Equatable {
    /// The `ETag` HEY sent with the body, sent back as `If-None-Match` on the next read.
    public var etag: String
    /// The body as HEY answered it.
    public var body: Data
    /// The headers HEY answered the body with, less any credential.
    public var headers: HTTPHeaders

    public init(etag: String, body: Data, headers: HTTPHeaders = HTTPHeaders()) {
        self.etag = etag
        self.body = body
        self.headers = headers
    }

    /// The cached headers, with each one a 304 carried replacing what was held under its name.
    func headersUpdated(by fresh: HTTPHeaders) -> HTTPHeaders {
        var merged = headers
        for name in fresh.names { merged.remove(name) }
        for (name, value) in fresh { merged.add(name, value) }
        return merged
    }
}

/// Stores JSON responses by `ETag`, so a repeated read can be answered from a 304.
public protocol ResponseCache: Sendable {
    /// What is held under `key`, if anything.
    func get(_ key: String) -> CachedResponse?
    /// Holds a response under `key`, replacing whatever was there.
    func set(_ key: String, _ response: CachedResponse)
    /// Forgets what is held under `key`.
    func invalidate(_ key: String)
    /// Forgets everything.
    func clear()
}

/// Keeps cached responses for the life of the process. What ``HeyConfig/enableCache`` turns on
/// when no cache of your own is handed over.
public final class InMemoryCache: ResponseCache, @unchecked Sendable {
    private let lock = NSLock()
    private var entries: [String: CachedResponse] = [:]

    public init() {}

    public func get(_ key: String) -> CachedResponse? {
        lock.lock()
        defer { lock.unlock() }
        return entries[key]
    }

    public func set(_ key: String, _ response: CachedResponse) {
        lock.lock()
        defer { lock.unlock() }
        entries[key] = response
    }

    public func invalidate(_ key: String) {
        lock.lock()
        defer { lock.unlock() }
        entries.removeValue(forKey: key)
    }

    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        entries.removeAll()
    }

    /// How many responses are held.
    public var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return entries.count
    }
}

/// The key a read is cached under: the URL and the credentials it went out with — every header
/// the auth strategy set, not only `Authorization` — hashed, so one identity's reading is never
/// answered to another and no token sits in a key.
func cacheKey(url: String, credentials: String) -> String {
    sha256Hex(Data("\(credentials)\n\(url)".utf8))
}
